frontend/catalyst/jax_tracer.py

```diff
@@ -1306,8 +1299,27 @@ def trace_function(
         return res_expanded_tracers, in_sig, out_sig
 
 
+def _get_shots(qnode) -> int:
+    """Extract shots from qnode, handling None case.
+
+    Args:
+        qnode: The quantum node containing shots information.
+
+    Returns:
+        int: Device shots value, 0 if None.
+    """
+    # When PennyLane allows dynamic shots, update tracing to accept dynamic shots too
+    # Use JAX-compatible conditional to handle potentially traced values
+    shots_value = qnode._shots.total_shots  # pylint: disable=protected-access
+    if shots_value is None:
+        shots = 0
+    else:
+        shots = shots_value
+    return shots
+
+
 @debug_logger
-def trace_quantum_function(
+def trace_quantum_function(  # pylint: disable=too-many-locals,too-many-statements,too-many-branches  # noqa: C901
```

Comment:
```suggestion
def trace_quantum_function(
```
These macros never worked actually :sweat: